code/modules/reagents/chemistry/machinery/reagentgrinder.dm

```diff
@@ -64,334 +48,475 @@
 		if(EXPLODE_LIGHT)
 			SSexplosions.low_mov_atom += beaker
 
-/obj/machinery/reagentgrinder/RefreshParts()
-	. = ..()
-	speed = 1
-	for(var/datum/stock_part/servo/servo in component_parts)
-		speed = servo.tier
+/obj/machinery/reagentgrinder/add_context(atom/source, list/context, obj/item/held_item, mob/user)
+	. = NONE
+	if(isnull(held_item))
+		if(!QDELETED(beaker) && !operating)
+			context[SCREENTIP_CONTEXT_RMB] = "Remove beaker"
+			. = CONTEXTUAL_SCREENTIP_SET
+		return
+
+	if(is_reagent_container(held_item) && held_item.is_open_container() && !operating)
+		if(QDELETED(beaker))
+			context[SCREENTIP_CONTEXT_LMB] = "Insert beaker"
+		else
+			context[SCREENTIP_CONTEXT_LMB] = "Replace beaker"
+		return CONTEXTUAL_SCREENTIP_SET
+
+	if(held_item.tool_behaviour == TOOL_SCREWDRIVER)
```

Comment:
this might be able to be a switch statement, although the additional check might have to be extracted out

Reply:
it would have been written as a switch if it wasn't for the crowbar act which requires to check for an open panel. Best to leave it at this